### DISCUSSION TITLE
Populated field type

### DISCUSSION THREAD

**Discussion by wasya243** [typegoose/typegoose#813] [Q&A] [answered]
Hello! Please help me. Here is the code.

```ts
import { prop, getModelForClass, Ref } from '@typegoose/typegoose';
import mongoose from 'mongoose';

export class Organization {
  @prop({required: true})
  name!: string;

  @prop({
    ref: () => User,
    foreignField: 'organizationId',
    localField: '_id'
  })
  public users: Ref<User>[] = [];
}
export class User {
  @prop({required: true})
  name!: string;

  @prop({
    ref: () => Organization, 
    required: true,
    foreignField: '_id',
    localField: 'organizationId',
    justOne: true
  })
  private organization!: Ref<Organization>;

  @prop({ required: true })
  public organizationId!: string;

}

const OrganizationModel = getModelForClass(Organization, {
  schemaOptions: {
    toJSON: { virtuals: true },
    toObject: { virtuals: true }
  }
});

const UserModel = getModelForClass(User, {
  schemaOptions: {
    toJSON: { virtuals: true },
    toObject: { virtuals: true }
  }
});
async function main() {
  try {
    await mongoose.connect('mongodb://localhost:27017/', {dbName: 'migration-to-typegoose'});

    const organization = await OrganizationModel.create({
      name: 'Acme Inc.'
    });

    const users = await Promise.all([
      UserModel.create({
        name: 'Vasyl',
        organizationId: organization._id.toString()
      }),
      UserModel.create({
        name: 'Misha',
        organizationId: organization._id.toString()
      })
    ]);

    const user = await UserModel.findOne().populate('organization');

    // here is TS error that name does not exist on organization
    console.log(user.organization.name)

    // possible solutions

    // 1) const user = await UserModel.findOne().populate<{organization: Organization}>('organization');
    // 2) same as above but define interface 
    // 3) typecasting to any
    // 5) if isDocument

    process.exit(0);
  } catch (err) {
    console.error(err);
    process.exit(1);
  }
}

main();
```

There are 2 entities: organization and user. Organization has many users. I do populate on user (populating organization). After that I try acccessing .name on populated organization field. I get TS error.

**Basically my question is there any other way to access fields on populated field without interfaces, typecasts, isDocument?**

**Comment by hasezoey** [comment]
> I get TS error.

i think the error is missing

> Basically my question is there any other way to access fields on populated field without interfaces, typecasts, isDocument?

no, you need to use one of the following:
- `isDocument` by either early-returning or only doing processing in the `if` block
- ~~typecasing with [`!` (non-null assertion operator)](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator)~~
- use some assertion (like `assert.ok(isDocument(doc.path))`, normal nodejs assert wont work because it does not have proper typing, see [typegoose's assertion for example](https://github.com/typegoose/typegoose/blob/f5b3d4b3d272cc03c36a95e7a746ee6449b3afcb/src/internal/utils.ts#L611-L615))
- ~~use some type-casting via the [`NonNullable` utility type](https://www.typescriptlang.org/docs/handbook/utility-types.html#nonnullabletype) (basically the same as `!` just longer)~~
- use manual casting of the type (not recommended) via `const yourPath = doc.path as YourClass` (not recommended)

**Reply by wasya243** [reply]
Thanks for your answer! I guess the error is not missing -> I wrote "TS error that name does not exist on organization" in the code example. Regarding typecasting with "!" operator or NonNullable utility type -> it's not going to work or I got you wrong. If I put "!" like this `user!.organization!.name` I still get the following error Property 'name' does not exist on type 'Ref<Organization>'. Property 'name' does not exist on type 'ObjectId'. So problem is not about possible undefined/null values, but about types.

**Reply by hasezoey** [reply, marked as answer]
i dont know why i wrote this, but yes, `!` and `NonNullable` do not help in this way, sorry about that.

so to your original question, your only options are:
- use `isDocument` via a early-return (via `if () return` or `assertion`) or only operate in that if-block
- use manual type-casting

**Reply by wasya243** [reply]
Yeap, cool!

**Reply by olawalejuwonm** [reply]
https://github.com/typegoose/typegoose/discussions/813#discussioncomment-7045428

**Comment by olawalejuwonm** [comment]
Is this possible without casting, can typegoose automatically set the type when I do .populate for the field being populated?

**Reply by hasezoey** [reply]
no typegoose does not do this automatically, to do this automatically typegoose would require to overwrite the mongoose populate type (which typegoose avoids doing)

but mongoose provides *some* way to overwrite the path via a generic: `.populate<{ field: PopulatedType }>(selector)` ([see mongoose documentation](https://mongoosejs.com/docs/typescript/populate.html))

also in addition `populate` modifies the document in-place, but typescript does not support in-place type changes (to my knowledge) so to to actually make the mongoose way useful, you would have to do `doc = doc.populate()` (reassigning the value)

**Comment by olawalejuwonm** [comment]
When I use Ref<User> it should at least give the user type. Or what's even
advantage of Ref over using ObjectID?

**Reply by hasezoey** [reply]
this is basically the question this discussion was originally about, see comment https://github.com/typegoose/typegoose/discussions/813#discussioncomment-5569352

TL;DR: `Ref` is a combination of the ref-type (objectid in most cases) and the class itself, and need to be filtered out after population, because population can fail

**Comment by olawalejuwonm** [comment]
Ok, how do I filter that out please?

**Reply by hasezoey** [reply]
see https://github.com/typegoose/typegoose/discussions/813#discussioncomment-5569352 and [Reference other Classes: Population](https://typegoose.github.io/typegoose/docs/guides/advanced/reference-other-classes#population)

**Reply by olawalejuwonm** [reply]
Thanks
